File: archive/c/c-plus-plus/duplicate-character-counter.cpp
#include <array>
#include <format>
#include <iostream>
#include <string_view>

[[noreturn]] void usage() {
    std::cout << "Usage: please provide a string\n";
    std::exit(1);
}

int main(int argc, char* argv[]) {
    if (argc != 2) usage();

    std::string_view input{argv[1]};
    if (input.empty()) usage();

    std::array<std::size_t, 256> counts{};

    for (const auto c : input) {
        counts[c]++;
    }

    bool any_duplicates = false;

    for (const auto c : input) {
        if (counts[c] > 1) {
            std::cout << std::format("{}: {}\n", c, counts[c]);
            counts[c] = 0;
            any_duplicates = true;
        }
    }

    if (!any_duplicates) {
        std::cout << "No duplicate characters\n";
    }
}

File: archive/t/typescript/duplicate-character-counter.ts
const error_msg: string = "Usage: please provide a string"
let str: string = (process.argv.length == 3) ? process.argv[2] : "";
if (str.length == 0) {
    console.log(error_msg);
    process.exit(1);
}

let hash_map: Map<string, number> = new Map<string, number>();
for (let i = 0; i < str.length; i++) {
    let char: string = str[i];
    let count: number = hash_map.get(char) || 0;
    hash_map.set(char, count + 1);
}
let has_duplicate: boolean = false;
hash_map.forEach((value, key) => {
    if (value > 1) {
        has_duplicate = true;
        console.log(`${key}: ${value}`);
    }
});

if (!has_duplicate) {
    console.log("No duplicate characters");
}
